board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c

```diff
@@ -87,6 +106,264 @@ int board_early_init_f(void)
+static int m4_upgrade_required(struct hash *fit_hash)
+{
+	uint8_t tee_hash[FIT_MAX_HASH_LEN];
+	int tee_hash_len = -1;
+
+	if (optee_get_tee_hash(tee_hash, &tee_hash_len, NULL))
+		return false;
+
+	if (tee_hash_len != fit_hash->len ||
+	    memcmp(tee_hash, fit_hash->value, tee_hash_len))
+		return true;
+
+	return false;
+}
+
+static int m4_check_hash(const void *fit, const int image_noffset,
+			 struct hash *fit_hash)
+{
+	int noffset = 0;
+	int i;
+
+	fdt_for_each_subnode(noffset, fit, image_noffset) {
+		const char *name = fit_get_name(fit, noffset, NULL);
+		if (!strncmp(name, FIT_HASH_NODENAME,
+			     strlen(FIT_HASH_NODENAME)))
+			goto process_m4_hash;
+	}
+
+	printf("M4: Can't get hash node\n");
+	return -EINVAL;
+
+process_m4_hash:
+
+	if (fit_image_hash_get_value(fit, noffset,
+				     &fit_hash->value, &fit_hash->len)) {
+		printf("M4: Can't get hash value property\n");
+		return -EINVAL;
+	}
+
+	printf("M4: FIT hash => ");
+	for (i = 0; i < fit_hash->len; i++)
+		printf("%x", fit_hash->value[i]);
+	printf("\n");
+
+	if (m4_upgrade_required(fit_hash) == false) {
+		printf("M4: upgrade not required\n");
+		return 0;
+	}
+
+	return 1;
+}
+
+static int m4_get_fit_node(const void *fit)
+{
+	int images_noffset;
+	const char *name;
+	int noffset;
+	int ndepth;
+	int count;
+
+	images_noffset = fdt_path_offset(fit, FIT_IMAGES_PATH);
+	if (images_noffset < 0) {
+		printf("M4: Can't find images parent node '%s' (%s)\n",
+		       FIT_IMAGES_PATH, fdt_strerror(images_noffset));
+		return -EINVAL;
+	}
+
+	for (ndepth = 0, count = 0,
+	     noffset = fdt_next_node(fit, images_noffset, &ndepth);
+			(noffset >= 0) && (ndepth > 0);
+			noffset = fdt_next_node(fit, noffset, &ndepth)) {
+
+		if (ndepth == 1) {
+			name = fit_get_name(fit, noffset, NULL);
+			if (!strncmp(name, "m4", 2))
+				return noffset;
+ 			count++;
+		}
+	}
+
+	printf("M4: Can't find the FPGA M4 node\n");
+
+	return -EINVAL;
+}
+
+static void m4_boot(struct spi_flash *flash)
+{
+	int hash_len = 0, tee_hash_len = 0, tee_payload_len = 0;
+	struct mtd_info *mtd = 	&flash->mtd;
+	uint8_t tee_hash[FIT_MAX_HASH_LEN];
+	uint8_t hash[FIT_MAX_HASH_LEN];
+	size_t len;
+
+	if (optee_get_tee_hash(tee_hash, &tee_hash_len, &tee_payload_len)) {
+		printf("M4: cant boot, TEE not accessible\n");
+		return;
+	}
+
+	/* clear all memory to prevent against possible attacks */
+	memset((void *) M4_BASE, 0x00, M4_SIZE);
+
+	/* read the image */
+	len = round_up(tee_payload_len, mtd->writesize);
+	if (spi_flash_read(flash, 0, len, (void *) M4_BASE)) {
+		printf("M4: Failed to read from flash, can't boot\n");
+		return;
+	}
+
+	printf("\tM4 Entry: 0x%x\n", M4_ENTRY(M4_BASE));
+	printf("\tM4 Sram : 0x%x\n", M4_BASE);
+	printf("\tM4 Size : 0x%x\n", tee_payload_len);
+
+	/* do not boot if the FPGA is corrupted (ie, sha different than TEE's) */
+	if (calculate_hash((const void *) M4_BASE, tee_payload_len, "sha256",
+			   hash, &hash_len)) {
+		printf("M4: cant boot, unsupported hash algorithm\n");
```

Comment:
Hm.. so at this point we've loaded the binary into memory and if we can't boot for some reason we return here and at 2 other points below.

But another process could find a way to `writel(M4_ENTRY(M4_BASE), SIM0_RBASE + 0x70);` and the M4 would boot the firmware we just said was bad.

Should we establish an error path here which does another `memset` to make sure we don't leave the binary in memory (would also use that path for the 2 returns below).

I might be over thinking this.. @ricardosalveti what do you think?

Reply:
Are we booting from memory or from QSPI here? Because at this point the image is already flashed in QSPI, so not much to do to protect that (unless we decide to erase it, which could work as well, since the next reboot will force a new flash to happen).

Reply:
@ricardosalveti @ldts so maybe create a standard error path which erases the QSPI on unrecoverable issues like this (and the few below it?) -- is there a danger of just erasing the QSPI over and over?